File: rust/src/value.rs
use crate::error::{NatsuzoraError, Result};
use serde_json::Value as JsonValue;
use std::collections::HashMap;

/// Integer range per spec: -9007199254740991 to 9007199254740991 (JavaScript safe integers)
pub const INTEGER_MIN: i64 = -9_007_199_254_740_991;
pub const INTEGER_MAX: i64 = 9_007_199_254_740_991;

/// Runtime value type for Natsuzora templates
#[derive(Debug, Clone, PartialEq)]
pub enum Value {
    Null,
    Bool(bool),
    Integer(i64),
    String(String),
    Array(Vec<Value>),
    Object(HashMap<String, Value>),
}

impl Value {
    /// Convert a JSON value to a Natsuzora Value
    pub fn from_json(json: JsonValue) -> Result<Self> {
        match json {
            JsonValue::Null => Ok(Value::Null),
            JsonValue::Bool(b) => Ok(Value::Bool(b)),
            JsonValue::Number(n) => {
                if let Some(i) = n.as_i64() {
                    if i < INTEGER_MIN || i > INTEGER_MAX {
                        return Err(NatsuzoraError::TypeError {
                            message: format!("Integer out of range: {}", i),
                        });
                    }
                    Ok(Value::Integer(i))
                } else if let Some(f) = n.as_f64() {
                    // Try to convert float to integer if it's a whole number
                    if f.fract() == 0.0 && f >= INTEGER_MIN as f64 && f <= INTEGER_MAX as f64 {
                        Ok(Value::Integer(f as i64))
                    } else {
                        Err(NatsuzoraError::TypeError {
                            message: format!("Floating point numbers are not supported: {}", f),
                        })
                    }
                } else {
                    Err(NatsuzoraError::TypeError {
                        message: "Invalid number".to_string(),
                    })
                }
            }
            JsonValue::String(s) => Ok(Value::String(s)),
            JsonValue::Array(arr) => {
                let values: Result<Vec<Value>> = arr.into_iter().map(Value::from_json).collect();
                Ok(Value::Array(values?))
            }
            JsonValue::Object(obj) => {
                let mut map = HashMap::new();
                for (k, v) in obj {
                    map.insert(k, Value::from_json(v)?);
                }
                Ok(Value::Object(map))
            }
        }
    }

    /// Check if the value is truthy per spec section 3.4
    /// Falsy values: false, null, 0, "", [], {}
    pub fn is_truthy(&self) -> bool {
        match self {
            Value::Null => false,
            Value::Bool(b) => *b,
            Value::Integer(n) => *n != 0,
            Value::String(s) => !s.is_empty(),
            Value::Array(arr) => !arr.is_empty(),
            Value::Object(obj) => !obj.is_empty(),
        }
    }

    /// Stringify the value per spec section 3.3
    /// Only String, Integer, and Null can be stringified
    pub fn stringify(&self) -> Result<String> {
        match self {
            Value::String(s) => Ok(s.clone()),
            Value::Integer(n) => {
                if *n < INTEGER_MIN || *n > INTEGER_MAX {
                    return Err(NatsuzoraError::TypeError {
                        message: format!("Integer out of range: {}", n),
                    });
                }
                Ok(n.to_string())
            }
            Value::Null => Ok(String::new()),
            Value::Bool(_) => Err(NatsuzoraError::TypeError {
                message: "Cannot stringify boolean value".to_string(),
            }),
            Value::Array(_) => Err(NatsuzoraError::TypeError {
                message: "Cannot stringify array".to_string(),
            }),
            Value::Object(_) => Err(NatsuzoraError::TypeError {
                message: "Cannot stringify object".to_string(),
            }),
        }
    }

    /// Ensure the value is an array and return a reference to it
    pub fn as_array(&self) -> Result<&Vec<Value>> {
        match self {
            Value::Array(arr) => Ok(arr),
            _ => Err(NatsuzoraError::TypeError {
                message: format!("Expected array, got {}", self.type_name()),
            }),
        }
    }

    /// Get the type name for error messages
    pub fn type_name(&self) -> &'static str {
        match self {
            Value::Null => "null",
            Value::Bool(_) => "boolean",
            Value::Integer(_) => "integer",
            Value::String(_) => "string",
            Value::Array(_) => "array",
            Value::Object(_) => "object",
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn test_truthy() {
        assert!(!Value::Null.is_truthy());
        assert!(!Value::Bool(false).is_truthy());
        assert!(Value::Bool(true).is_truthy());
        assert!(!Value::Integer(0).is_truthy());
        assert!(Value::Integer(1).is_truthy());
        assert!(Value::Integer(-1).is_truthy());
        assert!(!Value::String("".to_string()).is_truthy());
        assert!(Value::String("hello".to_string()).is_truthy());
        assert!(!Value::Array(vec![]).is_truthy());
        assert!(Value::Array(vec![Value::Integer(1)]).is_truthy());
        assert!(!Value::Object(HashMap::new()).is_truthy());
    }

    #[test]
    fn test_stringify() {
        assert_eq!(
            Value::String("hello".to_string()).stringify().unwrap(),
            "hello"
        );
        assert_eq!(Value::Integer(42).stringify().unwrap(), "42");
        assert_eq!(Value::Integer(-42).stringify().unwrap(), "-42");
        assert_eq!(Value::Integer(0).stringify().unwrap(), "0");
        assert_eq!(Value::Null.stringify().unwrap(), "");

        assert!(Value::Bool(true).stringify().is_err());
        assert!(Value::Array(vec![]).stringify().is_err());
        assert!(Value::Object(HashMap::new()).stringify().is_err());
    }

    #[test]
    fn test_from_json() {
        let value = Value::from_json(json!({"name": "test", "count": 42})).unwrap();
        if let Value::Object(obj) = value {
            assert_eq!(obj.get("name"), Some(&Value::String("test".to_string())));
            assert_eq!(obj.get("count"), Some(&Value::Integer(42)));
        } else {
            panic!("Expected Object");
        }
    }
}

File: typescript/src/value.ts
/**
 * Natsuzora Value Utilities
 *
 * Truthiness evaluation and stringification.
 */

import { TypeError, NullValueError, EmptyStringError } from "./errors.ts";

export const INTEGER_MIN = -9007199254740991;
export const INTEGER_MAX = 9007199254740991;

/**
 * Check if a value is truthy according to Natsuzora rules.
 *
 * Falsy values:
 * - false
 * - null
 * - 0
 * - "" (empty string)
 * - [] (empty array)
 * - {} (empty object)
 */
export function isTruthy(value: unknown): boolean {
  if (value === false || value === null || value === undefined) {
    return false;
  }

  if (typeof value === "number") {
    return value !== 0;
  }

  if (typeof value === "string") {
    return value.length > 0;
  }

  if (Array.isArray(value)) {
    return value.length > 0;
  }

  if (typeof value === "object") {
    return Object.keys(value).length > 0;
  }

  return true;
}

/**
 * Convert a value to a string according to Natsuzora rules.
 *
 * - String: returned as-is
 * - Integer: converted to string
 * - null/undefined: NullValueError (use stringifyNullable for null -> "")
 * - Boolean: ERROR
 * - Array/Object: ERROR
 */
export function stringify(value: unknown): string {
  if (typeof value === "string") {
    return value;
  }

  if (typeof value === "number") {
    if (!Number.isInteger(value)) {
      throw new TypeError(`Cannot stringify non-integer number: ${value}`);
    }
    if (value < INTEGER_MIN || value > INTEGER_MAX) {
      throw new TypeError(`Integer out of range: ${value}`);
    }
    return String(value);
  }

  if (value === null || value === undefined) {
    throw new NullValueError("Cannot stringify null value without '?' modifier");
  }

  if (typeof value === "boolean") {
    throw new TypeError("Cannot stringify boolean value");
  }

  if (Array.isArray(value)) {
    throw new TypeError("Cannot stringify array");
  }

  if (typeof value === "object") {
    throw new TypeError("Cannot stringify object");
  }

  throw new TypeError(`Cannot stringify value of type ${typeof value}`);
}

/**
 * Stringify with nullable modifier (null -> empty string).
 */
export function stringifyNullable(value: unknown): string {
  if (value === null || value === undefined) {
    return "";
  }
  return stringify(value);
}

/**
 * Stringify with required modifier (null -> NullValueError, "" -> EmptyStringError).
 */
export function stringifyRequired(value: unknown): string {
  if (value === null || value === undefined) {
    throw new NullValueError("Cannot stringify null value with '!' modifier");
  }
  if (value === "") {
    throw new EmptyStringError("Cannot stringify empty string with '!' modifier");
  }
  return stringify(value);
}

/**
 * Ensure a value is an array.
 */
export function ensureArray(value: unknown): unknown[] {
  if (!Array.isArray(value)) {
    throw new TypeError(`Expected array, got ${typeof value}`);
  }
  return value;
}

/**
 * Normalize data for use in templates.
 * - Convert float whole numbers to integers
 * - Convert object keys to strings
 * - Recursively normalize nested structures
 */
export function normalizeData(data: unknown): unknown {
  if (data === null || data === undefined) {
    return data;
  }

  if (typeof data === "number") {
    return normalizeNumber(data);
  }

  if (Array.isArray(data)) {
    return data.map((item) => normalizeData(item));
  }

  if (typeof data === "object") {
    const result: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(data)) {
      result[key] = normalizeData(value);
    }
    return result;
  }

  return data;
}

/**
 * Type guard to check if a value is a Record<string, unknown>.
 */
export function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

/**
 * Normalize data and assert it's a record.
 * Throws if the input is not an object.
 */
export function normalizeRootData(data: unknown): Record<string, unknown> {
  if (!isRecord(data)) {
    throw new TypeError("Root data must be an object");
  }
  return normalizeData(data) as Record<string, unknown>;
}

/**
 * Normalize bindings data.
 */
export function normalizeBindings(bindings: Record<string, unknown>): Record<string, unknown> {
  return normalizeData(bindings) as Record<string, unknown>;
}

function normalizeNumber(value: number): number {
  // Reject NaN and Infinity
  if (!Number.isFinite(value)) {
    throw new TypeError(`Invalid number: ${value}`);
  }

  // Number.isInteger returns true for whole-number floats like 3.0
  // In JavaScript, 3.0 === 3 and Number.isInteger(3.0) === true
  if (!Number.isInteger(value)) {
    throw new TypeError(`Floating point numbers are not supported: ${value}`);
  }

  // Check safe integer range
  if (value < INTEGER_MIN || value > INTEGER_MAX) {
    throw new TypeError(`Integer out of range: ${value}`);
  }

  return value;
}
